Cilium Daemon CrashLoopBackoff - panic: runtime error: invalid memory address or nil pointer dereference - related to dupplicated cilium endpoint

### Is there an existing issue for this?

- [X] I have searched the existing issues

### What happened?

Clilum is installed on Cluster as A CNI with Egressgateway feature enable and l7proxy also

Everything was running smoothly until we noticed that all cilium pods were in `crashLoopBackOff`.
a `rollout restart` changed nothing,

For the diag, the message `level=info msg="regenerating all endpoints" reason="one or more identities created or deleted"` we proceeded by elimination, deleting all **ciliumNetworkPolicy** and the **CiliumEgressGatewayPolicy**.
but the problem persisted.

Then we deleted all CiliumEndpoints(cep) and rollout restart, and the Cilium pods worked again.

So the cause was the: **cep**
But the root cause was related to the restoration of a namespace with velero: the restored backup contained CiliumEndpoint resources. So restoring them totally broke Cilium and therefore the cluster: nothing worked.


In other words, Cilium was unable to manage the  dupplicated ciliumEndpoints.

**Considering that Cilium as a CNI is a main present component, I didn't expect it to break completely for this reason and jeopardize the whole cluster**.

I would have preferred Cilium to warn and ignore the duplicated ciliumEndpoints but **continue to work**.  It's too essential to break for so little.


### Cilium Version

1.15.5

### Kernel Version

`5.15.0-107-generic #117-Ubuntu`

### Kubernetes Version

`v1.28.4`

### Regression

_No response_

### Sysdump

_No response_

### Relevant log output

```shell
time="2024-06-19T13:21:58Z" level=info msg="Policy imported via API, recalculating..." policyAddRequest=3bd087fa-bfe0-479b-b883-8bef5dd2a427 policyRevision=13 subsys=daemon
time="2024-06-19T13:21:58Z" level=info msg="NetworkPolicy successfully added" k8sApiVersion= k8sNetworkPolicyName=deny-all-except-dns subsys=k8s-watcher
time="2024-06-19T13:21:59Z" level=info msg="Removing stale endpoint interfaces" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="Skipping kvstore configuration" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="Restored router address from node_config" file=/var/run/cilium/state/globals/node_config.h ipv4=10.5.2.22 ipv6="<nil>" subsys=node
time="2024-06-19T13:21:59Z" level=info msg="Initializing node addressing" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="Initializing cluster-pool IPAM" subsys=ipam v4Prefix=10.5.2.0/24 v6Prefix="<nil>"
time="2024-06-19T13:21:59Z" level=info msg="Restoring endpoints..." subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="Endpoints restored" failed=0 restored=4 subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="Addressing information:" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="  Cluster-Name: kubelablan" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="  Cluster-ID: 0" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="  Local node-name: myServer.com" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="  Node-IPv6: <nil>" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="  External-Node IPv4: 10.55.33.142" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="  Internal-Node IPv4: 10.5.2.22" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="  IPv4 allocation prefix: 10.5.2.0/24" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="  Loopback IPv4: 169.254.42.1" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="  Local IPv4 addresses:" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="  - 10.5.2.22" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="  - 10.55.33.142" subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="Waiting until local node addressing before starting watchers depending on it" subsys=k8s-watcher
time="2024-06-19T13:21:59Z" level=info msg="Node updated" clusterName=kubelablan nodeName=mynodex.com subsys=nodemanager
time="2024-06-19T13:21:59Z" level=info msg="Creating or updating CiliumNode resource" node=myServer.com subsys=nodediscovery
time="2024-06-19T13:21:59Z" level=info msg="Adding local node to cluster" node=myServer.com subsys=nodediscovery
time="2024-06-19T13:21:59Z" level=info msg="Waiting until all pre-existing resources have been received" subsys=k8s-watcher
time="2024-06-19T13:21:59Z" level=info msg="Policy Add Request" ciliumNetworkPolicy="[&{EndpointSelector:{\"matchLabels\":{\"any:run\":\"test\",\"k8s:io.kubernetes.pod.namespace\":\"test\"}} NodeSelector:{} Ingress:[] IngressDeny:[] Egress:[{EgressCommonRule:{ToEndpoints:[] ToRequires:[] ToCIDR:[10.55.45.130/32] ToCIDRSet:[] ToEntities:[] ToServices:[] ToGroups:[] aggregatedSelectors:[{LabelSelector:0xc000fc8a60 requirements:0xc0012cccd8 cachedLabelSelectorString:&LabelSelector{MatchLabels:map[string]string{cidr.10.55.45.130/32: ,},MatchExpressions:[]LabelSelectorRequirement{},}}]} ToPorts:[] ToFQDNs:[] ICMPs:[] Authentication:<nil>} {EgressCommonRule:{ToEndpoints:[] ToRequires:[] ToCIDR: ToCIDRSet:[] ToEntities:[] ToServices:[] ToGroups:[] aggregatedSelectors:[{LabelSelector:0xc0008de7e0 requirements:0xc0005f28d0 cachedLabelSelectorString:&LabelSelector{MatchLabels:map[string]string{reserved.none: ,},MatchExpressions:[]LabelSelectorRequirement{},}}]} ToPorts:[] ToFQDNs:[{MatchName:kubedashboard.appli MatchPattern:}] ICMPs:[] Authentication:<nil>} {EgressCommonRule:{ToEndpoints:[] ToRequires:[] ToCIDR: ToCIDRSet:[] ToEntities:[] ToServices:[] ToGroups:[] aggregatedSelectors:[{LabelSelector:0xc0008de7e0 requirements:0xc0005f28d0 cachedLabelSelectorString:&LabelSelector{MatchLabels:map[string]string{reserved.none: ,},MatchExpressions:[]LabelSelectorRequirement{},}}]} ToPorts:[{Ports:[{Port:443 Protocol:TCP}] TerminatingTLS:<nil> OriginatingTLS:<nil> ServerNames:[] Listener:<nil> Rules:<nil>}] ToFQDNs:[{MatchName:www.google.fr MatchPattern:} {MatchName:www.adie.org MatchPattern:}] ICMPs:[] Authentication:<nil>}] EgressDeny:[] Labels:[k8s:io.cilium.k8s.policy.derived-from=CiliumNetworkPolicy k8s:io.cilium.k8s.policy.name=test-cnp k8s:io.cilium.k8s.policy.namespace=test k8s:io.cilium.k8s.policy.uid=7857e56f-8851-4c24-bb9e-a83ffd17a599] Description:}]" policyAddRequest=be13e500-08a2-4d54-8e2e-fb8d00e87908 subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="Policy imported via API, recalculating..." policyAddRequest=be13e500-08a2-4d54-8e2e-fb8d00e87908 policyRevision=14 subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="Imported CiliumNetworkPolicy" ciliumNetworkPolicyName=test-cnp k8sApiVersion=cilium.io/v2 k8sNamespace=test subsys=k8s-watcher
time="2024-06-19T13:21:59Z" level=info msg="Node updated" clusterName=kubelablan nodeName=myServerx.com subsys=nodemanager
time="2024-06-19T13:21:59Z" level=info msg="Node updated" clusterName=kubelablan nodeName=myServery.com subsys=nodemanager
time="2024-06-19T13:21:59Z" level=info msg="Node updated" clusterName=kubelablan nodeName=myServerz.com subsys=nodemanager
time="2024-06-19T13:21:59Z" level=info msg="Node updated" clusterName=kubelablan nodeName=myServerw.com subsys=nodemanager
time="2024-06-19T13:21:59Z" level=info msg="Node updated" clusterName=kubelablan nodeName=myServerv.com subsys=nodemanager
time="2024-06-19T13:21:59Z" level=info msg="Policy Add Request" ciliumNetworkPolicy="[&{EndpointSelector:{\"matchLabels\":{\"k8s:io.kubernetes.pod.namespace\":\"test\"}} NodeSelector:{} Ingress:[] IngressDeny:[] Egress:[{EgressCommonRule:{ToEndpoints:[{\"matchLabels\":{\"k8s:io.kubernetes.pod.namespace\":\"kube-system\",\"k8s:k8s-app\":\"kube-dns\"}}] ToRequires:[] ToCIDR: ToCIDRSet:[] ToEntities:[] ToServices:[] ToGroups:[] aggregatedSelectors:[]} ToPorts:[{Ports:[{Port:53 Protocol:ANY}] TerminatingTLS:<nil> OriginatingTLS:<nil> ServerNames:[] Listener:<nil> Rules:0xc0002541c0}] ToFQDNs:[] ICMPs:[] Authentication:<nil>}] EgressDeny:[] Labels:[k8s:io.cilium.k8s.policy.derived-from=CiliumNetworkPolicy k8s:io.cilium.k8s.policy.name=test-cnp-dns k8s:io.cilium.k8s.policy.namespace=test k8s:io.cilium.k8s.policy.uid=3f445b8e-427e-418f-99d9-2755e738b2fb] Description:}]" policyAddRequest=55004d75-9480-4b3b-aa59-75aa0df5e775 subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="Policy imported via API, recalculating..." policyAddRequest=55004d75-9480-4b3b-aa59-75aa0df5e775 policyRevision=15 subsys=daemon
time="2024-06-19T13:21:59Z" level=info msg="Imported CiliumNetworkPolicy" ciliumNetworkPolicyName=test-cnp-dns k8sApiVersion=cilium.io/v2 k8sNamespace=test subsys=k8s-watcher
time="2024-06-19T13:21:59Z" level=info msg="Initializing identity allocator" subsys=identity-cache
time="2024-06-19T13:21:59Z" level=info msg="Allocating identities between range" cluster-id=0 max=65535 min=256 subsys=identity-cache
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_host.forwarding sysParamValue=1
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_host.rp_filter sysParamValue=0
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_host.accept_local sysParamValue=1
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_host.send_redirects sysParamValue=0
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_net.forwarding sysParamValue=1
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_net.rp_filter sysParamValue=0
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_net.accept_local sysParamValue=1
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_net.send_redirects sysParamValue=0
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_vxlan.forwarding sysParamValue=1
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_vxlan.rp_filter sysParamValue=0
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_vxlan.accept_local sysParamValue=1
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.cilium_vxlan.send_redirects sysParamValue=0
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.core.bpf_jit_enable sysParamValue=1
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.conf.all.rp_filter sysParamValue=0
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=net.ipv4.fib_multipath_use_neigh sysParamValue=1
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=kernel.unprivileged_bpf_disabled sysParamValue=1
time="2024-06-19T13:21:59Z" level=info msg="Setting sysctl" subsys=sysctl sysParamName=kernel.timer_migration sysParamValue=0
time="2024-06-19T13:21:59Z" level=info msg="regenerating all endpoints" reason="one or more identities created or deleted" subsys=endpoint-manager
panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x0 pc=0x25c3346]

goroutine 260 [running]:
github.com/cilium/cilium/pkg/egressgateway.(*Manager).addEndpoint(0xc001626d80, 0xc001e3ee00)
        /go/src/github.com/cilium/cilium/pkg/egressgateway/manager.go:460 +0x286
github.com/cilium/cilium/pkg/egressgateway.(*Manager).handleEndpointEvent(0xc001a57ea8?, {{0x3944795, 0x6}, {{0xc001e960c0, 0x1e}, {0xc000b7a460, 0xc}}, 0xc001e3ee00, 0xc0010fe080})
        /go/src/github.com/cilium/cilium/pkg/egressgateway/manager.go:507 +0x5b
github.com/cilium/cilium/pkg/egressgateway.(*Manager).processEvents(0xc001626d80, {0x3f59438, 0xc00078c280})
        /go/src/github.com/cilium/cilium/pkg/egressgateway/manager.go:376 +0x425
created by github.com/cilium/cilium/pkg/egressgateway.newEgressGatewayManager.func
```


### Anything else?

_No response_

### Cilium Users Document

- [X] Are you a user of Cilium? Please add yourself to the [Users doc](https://github.com/cilium/cilium/blob/main/USERS.md)

### Code of Conduct

- [X] I agree to follow this project's Code of Conduct

--- a/pkg/egressgateway/endpoint.go
+++ b/pkg/egressgateway/endpoint.go
@@ -53,10 +53,6 @@ func getEndpointMetadata(endpoint *k8sTypes.CiliumEndpoint, identityLabels label
 		}
 	}
 
-	if endpoint.Identity == nil {
-		return nil, fmt.Errorf("endpoint has no identity metadata")
-	}
-
 	data := &endpointMetadata{
 		ips:    addrs,
 		labels: identityLabels.K8sStringMap(),

--- a/pkg/egressgateway/manager.go
+++ b/pkg/egressgateway/manager.go
@@ -434,6 +434,11 @@ func (manager *Manager) addEndpoint(endpoint *k8sTypes.CiliumEndpoint) error {
 		logfields.K8sUID:          endpoint.UID,
 	})
 
+	if endpoint.Identity == nil {
+		logger.Warning("Endpoint is missing identity metadata, skipping update to egress policy.")
+		return nil
+	}
+
 	if identityLabels, err = manager.getIdentityLabels(uint32(endpoint.Identity.ID)); err != nil {
 		logger.WithError(err).
 			Warning("Failed to get identity labels for endpoint")

